### DISCUSSION TITLE
bug: can not disable lap signature help in insert mode

### DISCUSSION THREAD

**Discussion by tranbinh1604** [LazyVim/LazyVim#3605] [Q&A] [answered]
### Did you check docs and existing issues?

- [X] I have read all the LazyVim docs
- [X] I have searched the existing issues of LazyVim
- [X] I have searched the existing issues of plugins related to this issue

### Neovim version (nvim -v)

NVIM v0.10.0 Build type: Release LuaJIT 2.1.1716656478

### Operating system/version

MacOS 14.5 (M2)

### Describe the bug

im not sure is it a bug but I want to disable the <C-k> while typing in insert mode but no luck, what I have tried to:

1. disable by using opts method

```
{
  "neovim/nvim-lspconfig",
  opts = function()
    local keys = require("lazyvim.plugins.lsp.keymaps").get()
    -- disable a keymap
    keys[#keys + 1] = { "K", false }
  end,
}
```

2. override by mapping <C-k> with other command in insert mode 
`map("i", "<C-k>", "<Up>", { desc = "move up" })`

3. return a empty table of keys 

```
-- LSP keymaps
{
  "neovim/nvim-lspconfig",
  keys = function ()
    return {}
  end,
}

```
4. disable  noice plugin completely 

all I need is set <C-h>, <C-j>, <C-k>, <C-l> to move the cursor in insert mode, but I have no luck with <C-k> insert mode, 

the command `:verbose imap <C-k>` alway return whatever I did 

`
i  <C-K>       *@<Lua 4122: /opt/homebrew/Cellar/neovim/0.10.0/share/nvim/runtime/lua/vim/lsp/buf.lua:115>
                 Signature Help
	Last set from Lua (run Nvim with -V1 for more details)
`

### Steps To Reproduce

1. install the latest lazyvim
2. set keymap in config/keymaps.lua
```map("i", "<C-b>", "<ESC>^i", { desc = "move beginning of line" })
map("i", "<C-e>", "<End>", { desc = "move end of line" })
map("i", "<C-h>", "<Left>", { desc = "move left" })
map("i", "<C-l>", "<Right>", { desc = "move right" })
map("i", "<C-j>", "<Down>", { desc = "move down" })
map("i", "<C-k>", "<Up>", { desc = "move up" })
```
3. open any file, go to insert mode, type anything then press C-k then noice will show "No signature help available" or a signature help window will show.

### Expected Behavior

The keymap <C-k> should be disable and the key binding <C-k> to up should work as it should be.

### Repro

```Lua
-- DO NOT change the paths and don't remove the colorscheme
local root = vim.fn.fnamemodify("./.repro", ":p")

-- set stdpaths to use .repro
for _, name in ipairs({ "config", "data", "state", "cache" }) do
  vim.env[("XDG_%s_HOME"):format(name:upper())] = root .. "/" .. name
end

-- bootstrap lazy
local lazypath = root .. "/plugins/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", lazypath, })
end
vim.opt.runtimepath:prepend(lazypath)

-- install plugins
local plugins = {
  "folke/tokyonight.nvim",
  "folke/LazyVim",
  -- add any other plugins here
}
require("lazy").setup(plugins, {
  root = root .. "/plugins",
})

vim.cmd.colorscheme("tokyonight")
-- add anything else here
```

**Comment by dpetka2001** [comment]
You're actually setting `K` to `false` in nvim-lspconfig keys when you actually want to set `<c-k>` to `false`. This is not a bug but a user configuration question and should rather be asked in the Discussions section instead.

**Reply by tranbinh1604** [reply]
sorry for the miss typed in previous post, the correct one is `keys[#keys + 1] = { "<C-k>", false }`, not just K
But the problem is even I set `C-k` to false in `lspconfig`, the lsp signature help still work when I’m in insert mode

![CleanShot 2024-06-12 at 19 37 05@2x](https://github.com/LazyVim/LazyVim/assets/162952043/15a7fd43-ab05-4d20-9016-fc879d000dd3)

However, when Im going to insert mode, press C-k, the signature help still popup 

![CleanShot 2024-06-12 at 19 45 26](https://github.com/LazyVim/LazyVim/assets/162952043/bf6cd287-57f6-4d70-b2cb-cf5c1e9b5559)

**Comment by tranbinh1604** [comment]
@dpetka2001 so wired

```
return {
  "neovim/nvim-lspconfig",
  opts = function(_, opts)
    local keys = require("lazyvim.plugins.lsp.keymaps").get()

    keys[#keys + 1] = { "<C-k>", false }  -- does not work, <C-k> still can active in insert mode
    keys[#keys + 1] = { "K", false } -- disabled as it should be
    keys[#keys + 1] = { "gK", vim.lsp.buf.hover, desc = "show hover" } -- changed as it should be

    opts.diagnostics = { -- set border for diagnostics floating
      float = {
        border = "rounded",
      },
    }
  end,
}
```

somehow Im not able to disable `<C-k>`, I also:
- search in keycaps list by `<leader>sk` and signature_help does not appear, 
- `:verbose imap <C-k>` still there 
![CleanShot 2024-06-12 at 22 47 55](https://github.com/LazyVim/LazyVim/assets/162952043/65b5a25e-3722-4a1e-9c92-f85b65b59491)

 but in insert mode still popup when press `<C-k>`, I also tried to delete all folders `.local/state/nvim`, `.local/share/nvim`, `.cache/nvim`, no luck at all

**Reply by folke** [reply, marked as answer]
you need to specify the modes you want to delete, so add `mode = "i"`, or `mode = {'i', 'n'}`

**Reply by tranbinh1604** [reply]
thanks you :) its work, seem the next 2 keys is in normal mode then does not need `mode = "i"`

**Reply by acdoussan** [reply]
sorry, im having the same issue, but this answer isn't making any sense to me. Is there a complete but contained example for the correct way to disable this? I have tried:

```
return {
  {
    "neovim/nvim-lspconfig",
    opts = function()
      local keys = require("lazyvim.plugins.lsp.keymaps").get()

      -- disable signature help in insert mode
       keys[#keys + 1] = { "<C-k>", false, { mode = { "i" } } }
    end,
  }
}
```

```
return {
  {
    "neovim/nvim-lspconfig",
    opts = function()
      local keys = require("lazyvim.plugins.lsp.keymaps").get()

      -- disable signature help in insert mode
       keys[#keys + 1] = { "i", "<C-k>", false }
    end,
  }
}
```

but neither of these work

**Reply by acdoussan** [reply]
Found some documentation and looked at the code, the following is a minimal example that worked for me. If you want to change a different keymap I'd recommend looking at the resources attached at the end.

```
return {
  {
    "neovim/nvim-lspconfig",
    opts = function()
      local keys = require("lazyvim.plugins.lsp.keymaps").get()

      -- disable signature help in insert mode
       keys[#keys + 1] = { "<c-k>", false, mode = "i" }
    end,
  }
}

```

references:
https://www.lazyvim.org/plugins/lsp#%EF%B8%8F-customizing-lsp-keymaps
https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/plugins/lsp/keymaps.lua#L24

**Reply by dpetka2001** [reply]
The OP wanted to change `<C-k>` in `insert` mode, so he could use it to move the cursor upwards. Do you want the same thing? If yes, the suggested answer should work. If no and you just want to disable the lsp signature help from automatically popping up, then you probably want the following
```lua
return {
  {
    "folke/noice.nvim",
    opts = {
      lsp = {
        signature = {
          auto_open = {
            enabled = false,
          },
        },
      },
    },
  },
}
```

**Reply by juanlopez4691** [reply]
This works for me:

```lua
local getKeys = function()
  local keys = require("lazyvim.plugins.lsp.keymaps").get()
  keys[#keys + 1] = { "<c-k>", false, mode = "i" }

  return keys
end

return {
  "neovim/nvim-lspconfig",
  opts = {
    keys = getKeys(),
    servers = {
      -- Servers config
    }
  }
}
```

**Comment by AdrianDiazCode** [comment]
I'm having this same problem. Nothing seems to work to remap c-k on insert mode

**Reply by Kyslik** [reply]
This should do it:

```lua
return {
  "neovim/nvim-lspconfig",
  opts = {
    servers = {
      ["*"] = {
        keys = {
          { "<C-k>", false, mode = "i" },
        },
      },
    },
  },
}
```

Ref: https://github.com/LazyVim/LazyVim/blob/c64a61734fc9d45470a72603395c02137802bc6f/lua/lazyvim/plugins/lsp/keymaps.lua#L11-L29
